### PR TITLE
Add platform filter to dependencies

### PR DESCRIPTION
**Pull request by raptorxcz**
Solve issue #900.

### CODE REVIEW

**Review by davidrothera — APPROVED** [review]

**Comment by raptorxcz** [comment]
Fixed conflict

**Comment by yonaskolb** [comment]
Thanks for your work her @raptorxcz. A couple of points:
- lets rename `platformFilter` to just `platform` everywhere
- perhaps `all` is a more suitable name than `macOS+iOS`? What do you think, I can be persuaded otherwise

**Comment by raptorxcz** [comment]
@yonaskolb thank for review. 

You have valid point. I updated PR.

**Comment by raptorxcz** [comment]
Fixed merge conflict

File: Sources/ProjectSpec/Dependency.swift [hunk 1]
```diff
@@ -31,6 +33,17 @@ public struct Dependency: Equatable {
...
         self.weakLink = weakLink
+        self.platform = platform
+    }
+    
+    public enum Platform: Equatable {
+        case all
+        case iOS
+        case mac
+        
+        public static func == (lhs: Self, rhs: Self) -> Bool {
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Dependency.swift:44]: This equatable function isn't required, and in this case actually returns the wrong result

File: Sources/ProjectSpec/Dependency.swift [hunk 2]
```diff
@@ -31,6 +33,17 @@ public struct Dependency: Equatable {
         self.link = link
         self.implicit = implicit
         self.weakLink = weakLink
+        self.platform = platform
+    }
+    
+    public enum Platform: Equatable {
+        case all
+        case iOS
+        case mac
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Dependency.swift:42]: Should we use `macOS` to match the iOS and the platform names from the rest of Xcodegen

File: Sources/XcodeGenKit/PBXProjGenerator.swift [hunk 3]
```diff
@@ -1299,6 +1304,17 @@ public class PBXProjGenerator {
             targetObject.productType = target.type
         }
     }
+    
+    private func makePlatfor(for platform: Dependency.Platform) -> String? {
```
Comment (yonaskolb) [inline Sources/XcodeGenKit/PBXProjGenerator.swift:1308]: Platfor

File: Tests/Fixtures/TestProject/project.yml [hunk 4]
```diff
@@ -113,12 +113,16 @@ targets:
       PRODUCT_BUNDLE_IDENTIFIER: com.project.app
     dependencies:
       - target: Framework_iOS
+        platform: mac+iOS
```
Comment (yonaskolb) [inline Tests/Fixtures/TestProject/project.yml:116]:
```suggestion
        platform: all
```

File: Tests/Fixtures/TestProject/project.yml [hunk 5]
```diff
@@ -113,12 +113,16 @@ targets:
...
       - target: StaticLibrary_ObjC_iOS
       - carthage: Result
+        platform: mac
```
Comment (yonaskolb) [inline Tests/Fixtures/TestProject/project.yml:119]: if we change to macOS
```suggestion
        platform: macOS
```

File: Tests/ProjectSpecTests/SpecLoadingTests.swift [hunk 6]
```diff
@@ -376,9 +376,9 @@ class SpecLoadingTests: XCTestCase {
             $0.it("parses target dependencies") {
                 var targetDictionary = validTarget
                 targetDictionary["dependencies"] = [
-                    ["target": "name", "embed": false],
-                    ["target": "project/name", "embed": false],
-                    ["carthage": "name", "findFrameworks": true],
+                    ["target": "name", "embed": false, "platform": "all"],
+                    ["target": "project/name", "embed": false, "platform": "mac"],
```
Comment (yonaskolb) [inline Tests/ProjectSpecTests/SpecLoadingTests.swift:380]: If we change to macOS
```suggestion
                    ["target": "project/name", "embed": false, "platform": "macOS"],
```

File: Sources/ProjectSpec/Dependency.swift (hunk 2 above)
Comment (raptorxcz) [reply Sources/ProjectSpec/Dependency.swift]: Updated in dfc471d8511f37738bfb9a84915d822bcc299f1e.

File: Sources/XcodeGenKit/PBXProjGenerator.swift (hunk 3 above)
Comment (raptorxcz) [reply Sources/XcodeGenKit/PBXProjGenerator.swift]: Updated in dfc471d8511f37738bfb9a84915d822bcc299f1e.

File: Tests/Fixtures/TestProject/project.yml (hunk 4 above)
Comment (raptorxcz) [reply Tests/Fixtures/TestProject/project.yml]: Updated in dfc471d8511f37738bfb9a84915d822bcc299f1e.

File: Tests/Fixtures/TestProject/project.yml (hunk 5 above)
Comment (raptorxcz) [reply Tests/Fixtures/TestProject/project.yml]: Updated in dfc471d8511f37738bfb9a84915d822bcc299f1e.

File: Tests/ProjectSpecTests/SpecLoadingTests.swift (hunk 6 above)
Comment (raptorxcz) [reply Tests/ProjectSpecTests/SpecLoadingTests.swift]: Updated in dfc471d8511f37738bfb9a84915d822bcc299f1e.

File: Sources/ProjectSpec/Dependency.swift (hunk 1 above)
Comment (yonaskolb) [reply Sources/ProjectSpec/Dependency.swift:44]: If you make Platform a RawRepresentable `Platform: String, Equatable` Swift can auto generate the Equatable conformance. This also makes the decoding code simpler as you can init with a string

Comment (raptorxcz) [reply Sources/ProjectSpec/Dependency.swift]: I refactored dependence equatable in a1ebf2240158f49f99a71ed3cac5a4a5c60308e4.

File: Tests/Fixtures/TestProject/Project.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -507,6 +510,7 @@
 			files = (
 				535A98A3E3B74E09891D977F /* TestFramework.framework in Embed Frameworks */,
 				BD1419893577E6CEDF8CBA83 /* Result.framework in Embed Frameworks */,
+				BB06A57E259D0D2A001EA21F /* Result.framework in Embed Frameworks */,
```
Comment (raptorxcz) [inline Tests/Fixtures/TestProject/Project.xcodeproj/project.pbxproj:513]: As you can see, new copy of Result dependency is added. This is result of fixing Platforms equatable. I tried to understand this behaviour, but I failed. I do not even know if is error or not.

**Comment by jlott1** [comment]
> Thanks for your work her @raptorxcz. A couple of points:
> 
> * lets rename `platformFilter` to just `platforms` everywhere
> * perhaps `all` is a more suitable name than `macOS+iOS`, and easier to type? What do you think, I can be persuaded otherwise

@yonaskolb 
I've got a question.   Could this also be used to have platform specific dependencies, for example I need to install a framework dependency only for iOS but not include it for tvOS.   Is this possible with this change or can it be made possible?

**Comment by yonaskolb** [comment]
That's a good question @jlott1. The work done here doesn't achieve that, this is more aimed at Catalyst, is that right @raptorxcz?
Being able to only add dependencies for a specific platform would be useful though for cross platform targets. I worry that people will think this is what that is. Thoughts @raptorxcz?

**Comment by yonaskolb** [comment]
It would be great to have some unit tests around the generation of this as well

**Comment by raptorxcz** [comment]
This PR add only support for platform option at Xcode.
Screen Shot 2021-02-15 at 9 16 26

**Comment by raptorxcz** [comment]
Added some tests in 24e1ca0882a8b9a2000bfd14cfd065b207d0ddc6.

**Comment by raptorxcz** [comment]
@yonaskolb Could you look at my PR?

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,6 +10,7 @@
 - Added `basedOnDependencyAnalysis` to Project Spec Build Script to be able to choose not to skip the script. [#992](https://github.com/yonaskolb/XcodeGen/pull/992) @myihsan
 - Add `BuildRule.runOncePerArchitecture` to allow running build rules once per architecture. [#950](https://github.com/yonaskolb/XcodeGen/pull/950) @sascha
 - Adds discovered dependency file for a build script [#1012](https://github.com/yonaskolb/XcodeGen/pull/1012) @polac24 @fggeraissate
+- Added ability to set custom platform for dependency [#934](https://github.com/yonaskolb/XcodeGen/pull/934) @raptorxcz
```
Comment (freddi-kit) [inline CHANGELOG.md:13]: Please move it to `#Next Version` when solving conflicts

File: Sources/ProjectSpec/Dependency.swift [hunk 9]
```diff
@@ -22,7 +23,8 @@ public struct Dependency: Equatable {
         codeSign: Bool? = nil,
         link: Bool? = nil,
         implicit: Bool = implicitDefault,
-        weakLink: Bool = weakLinkDefault
+        weakLink: Bool = weakLinkDefault,
+        platform: Platform = .all
```
Comment (freddi-kit) [inline Sources/ProjectSpec/Dependency.swift:27]: The default value is defined in top of this struct like below code

https://github.com/yonaskolb/XcodeGen/blob/24e1ca0882a8b9a2000bfd14cfd065b207d0ddc6/Sources/ProjectSpec/Dependency.swift#L5-L7

Could you help add like `platformDefault`?

File: Sources/ProjectSpec/Dependency.swift [hunk 10]
```diff
@@ -31,6 +33,13 @@ public struct Dependency: Equatable {
         self.link = link
         self.implicit = implicit
         self.weakLink = weakLink
+        self.platform = platform
```
Comment (freddi-kit) [inline Sources/ProjectSpec/Dependency.swift:36]: Please apply `platformDefault` which mentioned above

File: Docs/ProjectSpec.md [hunk 11]
```diff
@@ -428,6 +428,7 @@ A dependency can be one of a 6 types:
 - [ ] **codeSign**: **Bool** - Whether the `codeSignOnCopy` setting is applied when embedding framework. Defaults to true
 - [ ] **removeHeaders**: **Bool** - Whether the `removeHeadersOnCopy` setting is applied when embedding the framework. Defaults to true
 - [ ] **weak**: **Bool** - Whether the `Weak` setting is applied when linking the framework. Defaults to false
+- [ ] **platform**: **String** - Add dependency to selected platforms. Available platforms are: **iOS**, **macOS** and **all**. Defaults is **all**.
```
Comment (freddi-kit) [inline Docs/ProjectSpec.md:431]:
```suggestion
- [ ] **platform**: **String** - Add dependency to selected platforms. Available platforms are: **iOS**, **macOS** and **all**. Defaults is **all**
```

File: Sources/ProjectSpec/Dependency.swift (hunk 10 above)
Comment (raptorxcz) [reply Sources/ProjectSpec/Dependency.swift:37]: Updated in 07df57d86dfd3172a9e49793ab04a9eb76e7414d.

File: Sources/ProjectSpec/Dependency.swift (hunk 9 above)
Comment (raptorxcz) [reply Sources/ProjectSpec/Dependency.swift]: Updated in 07df57d86dfd3172a9e49793ab04a9eb76e7414d.

File: CHANGELOG.md (hunk 8 above)
Comment (raptorxcz) [reply CHANGELOG.md]: Updated in 07df57d86dfd3172a9e49793ab04a9eb76e7414d.

**Comment by raptorxcz** [comment]
Could you look at my PR? Thanks

**Comment by yonaskolb** [comment]
Thanks @raptorxcz. Sorry for the delay on this

**Comment by daltonclaybrook** [comment]
> That's a good question @jlott1. The work done here doesn't achieve that, this is more aimed at Catalyst, is that right @raptorxcz?
> Being able to only add dependencies for a specific platform would be useful though for cross platform targets. I worry that people will think this is what that is. Thoughts @raptorxcz?

@yonaskolb I am one such person who has found myself here because I thought this referred to platform-specific dependencies as mentioned by @jlott1. This is definitely a feature my team could benefit from. Is there a world where both of these features could exist? What would the spec need to look like to accommodate both?

**Pull request merged**